pkg/server/heartbeat.go

```diff
@@ -0,0 +1,142 @@
+package server
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/PagerDuty/go-pdagent/pkg/common"
+	"github.com/google/uuid"
+	"go.uber.org/zap"
+)
+
+const HEARTBEAT_URL = "https://api.pagerduty.com/agent/2014-03-14/heartbeat/go-pdagent"
+const HEARTBEAT_FREQUENCY_SECONDS = 60 * 60 // Send heartbeat every hour
+const HEARTBEAT_MAX_RETRIES = 10
+const RETRY_GAP_SECONDS = 10
+
+type HeartbeatTask struct {
+	heartbeatId        string
+	ticker             *time.Ticker
+	shutdown           chan bool
+	logger             *zap.SugaredLogger
+	client             *http.Client
+	heartbeatFrequency int
+}
+
+type HeartbeatResponseBody struct {
+	HeartBeatIntervalSeconds int `json:"heartbeat_interval_secs"`
+}
+
+func NewHeartbeatTask() *HeartbeatTask {
+	hb := HeartbeatTask{
+		heartbeatId:        uuid.NewString(),
+		ticker:             nil,
+		shutdown:           make(chan bool),
+		logger:             common.Logger.Named("Heartbeat"),
+		client:             &http.Client{},
+		heartbeatFrequency: HEARTBEAT_FREQUENCY_SECONDS,
+	}
+
+	return &hb
+}
+
+func (hb *HeartbeatTask) Start() {
+	hb.logger.Info("Starting heartbeat")
+	hb.ticker = time.NewTicker(time.Duration(hb.heartbeatFrequency) * time.Second)
+
+	go func() {
+		for {
+			select {
+			case <-hb.shutdown:
+				return
+			case <-hb.ticker.C:
+				go hb.beat()
+			}
+		}
+	}()
+}
+
+func (hb *HeartbeatTask) Shutdown() {
+	hb.ticker.Stop()
+	hb.shutdown <- true
+	hb.logger.Info("Heartbeat stopped")
+}
+
+func (hb *HeartbeatTask) beat() {
+	hb.logger.Info("Sending heartbeat")
+
+	attempts := 0
+
+	for {
+		attempts++
+
+		statusCode, isError := hb.makeHeartbeatRequest()
+		if isError {
+			return
+		}
+
+		if statusCode/100 == 2 {
+			hb.logger.Info("Heartbeat successful!")
+			return
+		} else if statusCode/100 == 5 {
+			hb.logger.Error("Error sending heartbeat - will retry")
+		} else {
+			hb.logger.Info("Heartbeat request returned a non-success response code - will retry")
+		}
+
+		if attempts >= HEARTBEAT_MAX_RETRIES {
+			hb.logger.Info("Heartbeat retry limit exceeded - will not retry")
+			return
+		}
+
+		hb.logger.Info("Sleeping before retry")
+		time.Sleep(RETRY_GAP_SECONDS * time.Second)
+		hb.logger.Info("Retrying heartbeat")
+	}
+}
+
+func (hb *HeartbeatTask) makeHeartbeatRequest() (int, bool) {
+	req, err := http.NewRequest("GET", HEARTBEAT_URL, nil)
+	if err != nil {
+		hb.logger.Error("Failed to create heartbeat request - will not retry")
+		return 0, true
+	}
+
+	req.Header.Add("User-Agent", userAgent(*hb))
+	req.Header.Add("Content-Type", "application/json")
+
+	httpResp, err := hb.client.Do(req)
+	if err != nil {
+		hb.logger.Error("Failed to send heartbeat request - will not retry")
```

Comment:
We probably do not need to log details about whether we are retrying here in the inner function.  This logging seems to be taken care of in the outer function.

Reply:
Removed the error logging from this function: 600c6bf